compatibility.bs

```diff
@@ -832,6 +846,142 @@ supported on all {{Window}} objects and <code><{body}></code> elements as attrib
+- It generates a [=permission key=] independently of the [=current settings object=].</div>
+
+</div>
+
+<h3 id="sa-heuristics-popup">Popup Heuristic</h3>
+
+Append the following steps to the [[!HTML]]'s <a spec="html">activation notification</a> algorithm:
+
+<div algorithm="user-activation-patch">
+
+1. Run [=detect a popup heuristic=] given <var ignore>document</var>.
+
+</div>
+
+<div algorithm>
+
+To <dfn>detect a popup heuristic</dfn> given a [=Document=] |document|,
+perform the following steps:
+
+1. Let |browsingContext| be |document|'s [=Document/browsing context=].
+1. If |browsingContext|'s [=browsing context/is popup=] is false, then return.
+1. If |browsingContext|'s [=browsing context/opener origin at creation=] is null, then return.
+1. Let |navigable| be |document|'s [=node navigable=].
+1. If |navigable| is null, then return.
+1. Let |topDocument| be |navigable|'s [=navigable/top-level traversable=]'s
+    [=navigable/active document=].
+1. Let |origin| be |topDocument|'s [=Document/origin=].
+1. If |origin| is an [=opaque origin=] then return.
+1. Let |site| be the result of running [=obtain a site=] given |origin|.
+1. Let |host| be |site|'s [=host=].
+1. [=Grant access for heuristics=] given:
+    <dl>
+        <dt><var ignore>host</var></dt>
+        <dd>|host|</dd>
+        <dt><var ignore>firstPartyHost</var></dt>
+        <dd>|browsingContext|'s <var ignore>opener origin at creation=</var></dd>
+        <dt><var ignore>duration</var></dt>
+        <dd>[=popup heuristic grant duration=]</dd>
+    </dl>
+
+</div>
+
+<h3 id="sa-heuristics-redirect">Redirect Heuristic</h3>
+
+Insert the following steps in [[!HTML]]'s <a spec="html">load a document</a> algorithm, before the
+final step which returns null.
+
+<div algorithm="loading-a-document-patch">
+
+1. Run [=detect a redirect heuristic=] given <var ignore>navigable</var>.
+
+</div>
+
+<div algorithm>
+
+To <dfn>detect a redirect heuristic</dfn> given a [=/navigable=] |navigable|, perform the following
+steps:
+
+1. Let |topDocument| be |navigable|'s [=navigable/top-level traversable=]'s
+    [=navigable/active document=].
+1. Let |firstPartyOrigin| be |topDocument|'s [=Document/origin=].
+1. If |firstPartyOrigin| is an [=opaque origin=] then return.
+1. Let |firstPartySite| be the result of running [=obtain a site=] given |firstPartyOrigin|.
+1. Let |firstPartyHost| be |firstPartySite|'s [=host=].
+1. Let |bounceTrackingRecord| be |navigable|'s [=navigable/top-level traversable=]'s [=top-level traversable/bounce tracking record=].
+1. [=list/For each=] |bounceUrl| in |bounceTrackingRecord|'s [=bounce tracking record/bounce set=]:
+  1. Let |bounceSite| be the result of running [=obtain a site=] given |bounceUrl|.
+  1. Let |bounceHost| be |bounceSite|'s [=host=].
+  1. If |bounceHost| [=host/equals=] |firstPartyHost|, [=iteration/continue=].
+  <!-- TODO: check if |bounceUrl| has a transient activation. This will require patching bounce tracking record. -->
```

Comment:
Is the idea to use [this map](https://privacycg.github.io/nav-tracking-mitigations/#user-activation-map) to do this checking?

Reply:
That user activation map won't suffice unfortunately, since it stores all historical interactions. Chrome's impl specifically checks not just for any recent interaction, but for an interaction during the last (extended) navigation.

We could filter the user activation map for *very* recent interactions but that approach is a bit hacky. Better to update the bounce tracking record to record sites with an interaction as the navigation is ongoing - that matches our code too.

Reply:
Done - updated the navtracking spec and referencing the new "user activation set".